components/ws-proxy/pkg/sshproxy/forward.go

```diff
@@ -0,0 +1,189 @@
+// Copyright (c) 2021 Gitpod GmbH. All rights reserved.
+// Licensed under the GNU Affero General Public License (AGPL).
+// See License-AGPL.txt in the project root for license information.
+
+package sshproxy
+
+import (
+	"fmt"
+	"io"
+	"net"
+	"sync"
+	"time"
+
+	"golang.org/x/crypto/ssh"
+)
+
+const GITPOD_USERNAME = "gitpod"
+
+func proxy(reqs1, reqs2 <-chan *ssh.Request, channel1, channel2 ssh.Channel) {
+	var closer sync.Once
+	closeFunc := func() {
+		channel1.Close()
+		channel2.Close()
+	}
+
+	defer closer.Do(closeFunc)
+
+	closerChan := make(chan bool, 1)
+
+	go func() {
```

Comment:
One of the two Go routines might leak because it tries to write to the channel with a capacity of one, but the reader might be gone at this point. Increasing the capacity to `2` would solve this, but also require a comment explaining why the channel has that capacity (see https://github.com/uber-go/guide/blob/master/style.md#channel-size-is-one-or-none).